### PR TITLE
Fix for PAM_USER env and more

### PR DESCRIPTION
**Pull request by kiavash-at-work**
This PR contains:

1. Fix for bug #7 
   * **NOTE**: The associated unit test is still not able to test for this bug and possible regression can happen.
2. Refactored code to make `main()` only for entry and exit. This helps in integration test implementation.
3. Added `Option<>` support to `Sargs` and related codes.
4. Refactored key calculation into a separate function to eliminate repeated codes.

### CODE REVIEW

File: shavee-bin/src/args.rs [hunk 1]
```diff
@@ -170,24 +162,42 @@ impl Sargs {
...
-                String::from("pam")
+                let user = match env::var("PAM_USER") {
+                    Ok(u) => u,
+                    Err(e) => {
+                        // FIX for "[bug] Command line parse error #7"
+                        // If PAM mode is selected but PAM_USER env doesn't exists
+                        // then return Err(error message)
+                        let error_message = e.to_string().to_owned();
+                        return Err(
+                            clap::Error::with_description(
```
Comment (kiavash-at-work) [inline shavee-bin/src/args.rs:173]: For `PAM_USER` env error message, this code converts the error message to a new `clap` message with `EmptyValue` error kind and then returns it to the calling function.

File: shavee-bin/src/args.rs [hunk 2]
```diff
@@ -128,35 +130,25 @@ impl Sargs {
...
 
         // if zset arg is entered, then its value will be used
-        // if not, then pass "NULL"
-        // NOTE: validating dataset is done in zfs.rs module
-        let mut dataset = match arg.value_of("zset") {
-            Some(s) => {
-                let mut s = s.to_string();
-                // remove the trailing '/' from dataset if entered
+        // NOTE: validating dataset is done by zfs module
+        let mut dataset = match arg.value_of("zset")
```
Comment (kiavash-at-work) [inline shavee-bin/src/args.rs:138]: Compare this implementation to line 179. Both can be done to manipulate contents wrap with `Option` and it depends on the programmer which one is preferred.

Here it is an example using `match`

File: shavee-bin/src/args.rs [hunk 3]
```diff
@@ -170,24 +162,42 @@ impl Sargs {
...
+                        let error_message = e.to_string().to_owned();
+                        return Err(
+                            clap::Error::with_description(
+                                &error_message[..], clap::ErrorKind::EmptyValue))
+                        },
+                    };
+                // append the value from "PAM_USER" to the end of dataset, if it exists
+                // otherwise keep None
+                dataset = dataset
+                    .map(|d| {
```
Comment (kiavash-at-work) [inline shavee-bin/src/args.rs:180]: Like the comment on line 138, here is an example using `map`.

Remember that this code is `closure` and [*captures*](https://doc.rust-lang.org/rust-by-example/fn/closures/capture.html) the environment variables.

File: shavee-bin/src/main.rs [hunk 4]
```diff
@@ -1,90 +1,54 @@
...
-                }
-            },
-            "create" => match create_zfs_yubi(pass, args.dataset, args.yslot) {
-                Ok(()) => exit(0),
-                Err(e) => {
-                    eprintln!("Error: {}", e);
-                    exit(1)
-                }
-            },
-            _ => unreachable!(),
```
Comment (kiavash-at-work) [inline shavee-bin/src/main.rs:46]: Using `enum` eliminates the need for "capture the rest" branch.

File: shavee-bin/src/main.rs [hunk 5]
```diff
@@ -1,90 +1,54 @@
...
         }
-    };
+    });
+}
+
+
+fn run(args: Sargs) -> Result<(), Box<dyn Error>> {
+
+    let pass = rpassword::prompt_password_stderr("Dataset Password: ")
+        .map_err(|e| e.to_string())?;
```
Comment (kiavash-at-work) [inline shavee-bin/src/main.rs:31]: Just like `map` that operates on `Ok(i)`, `map_err` operates on `Err(e)`. 

Here we need to convert the error to `String` which can be converted to `Box<dyn Error>`

File: shavee-bin/src/main.rs [hunk 6]
```diff
@@ -1,90 +1,54 @@
...
-            _ => unreachable!(),
+        Umode::File => match args.mode {
+            Mode::Print => print_mode_file(pass, args.file, args.port)?,
+            Mode::Mount => unlock_zfs_file(pass, args.file, args.dataset, args.port)?,
+            Mode::Create => create_zfs_file(pass, args.file, args.dataset, args.port)?,
         },
-        "password" => {
-            let key = hash_argon2(pass.into_bytes()).unwrap();
+        Umode::Password => {
+            let key =  hash_argon2(pass.into_bytes())?;
```
Comment (kiavash-at-work) [inline shavee-bin/src/main.rs:45]: This is a repeat of `file_key_calculation()` and instead it is possible to call `file_key_calculation()` this way:

```Rust
let key =  file_key_calculation(pass, None, None)?;
```

However if the above is implemented, `file_key_calculation()` must be changed to
* If `file` is `None` then doesn't call `get_filehash()` nor concatenation trait.

**Comment by ashuio** [comment]
Well all this changes Look good to me, Thanks for this PR. 

I'll merge it right away

**Pull request merged**